src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs

```diff
@@ -574,6 +574,39 @@ public async Task SendRecvPollSync_TcpListener_Socket(IPAddress listenAt, bool p
             }
         }
 
+        [Fact]
+        public async Task Send_0ByteSend_Success()
+        {
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(1);
+
+                Task<Socket> acceptTask = AcceptAsync(listener);
+                await Task.WhenAll(
+                    acceptTask,
+                    ConnectAsync(client, new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port)));
+
+                using (Socket server = await acceptTask)
+                {
+                    for (int i = 0; i < 3; i++)
+                    {
+                        // Zero byte send should be a no-op 
+                        int bytesSent = await SendAsync(client, new ArraySegment<byte>(Array.Empty<byte>()));
+                        Assert.Equal(0, bytesSent);
+
+                        // Socket should still be usable
+                        await SendAsync(client, new byte[] { 99 });
+                        byte[] buffer = new byte[10];
+                        int bytesReceived = await ReceiveAsync(server, buffer);
+                        Assert.Equal(1, bytesReceived);
+                        Assert.Equal(99, buffer[0]);
+                    }
+                }
+            }
+        }
```

Comment:
Nit: it might be worth to dedupe the test code by moving it to a helper method, that can be then invoked from the actual test cases:

```C#
protected Task<int> RunZeroByteSendTest(Action<Socket> doSendZeroBytes) { ... }
```

Even `SendReceive_SpanSyncForceNonBlocking` could do 
```C#
RunZeroByteSendTest(client => {
    Task<int> sendTask = client.SendAsync(ReadOnlyMemory<byte>.Empty, SocketFlags.None).AsTask();
    Assert.True(sendTask.IsCompleted);
    return sendTask;
});
```